### DISCUSSION TITLE
How to test code using async/await and MainActor.run

### DISCUSSION THREAD

**Discussion by kaylanx** [pointfreeco/swift-snapshot-testing#669] [Q&A]
We have been trying out the swift-snapshot-testing framework with little success.  Basically the snapshots are being taken before our state has updated.  

I've created an example for the problem we are facing... We we have a SwiftUI, that displays a list of numbers, those numbers are generated asynchronously, there is a loading state which will be initial, loading or finished, the updates to these are done using MainActor.run

```
import SwiftUI

struct NumbersView: View {
    enum LoadingState {
        case initial, loading, finished([Int])
    }

    final class ViewModel: ObservableObject {
        @Published var loadingState: LoadingState = .initial
        private var loadingInts = [Int]()

        func load() async {
            await MainActor.run { [weak self] in
                self?.loadingState = .loading
            }

            for x in 0..<10 {
                loadingInts.append(x)
            }

            await MainActor.run { [weak self] in
                self?.loadingState = .finished(loadingInts)
            }
        }
    }

    @ObservedObject var viewModel = ViewModel()

    var body: some View {
        VStack {
            if case .finished(let numbers) = viewModel.loadingState {
                List {
                    ForEach(numbers, id: \.self) { number in
                        Text("\(number)")
                    }
                }
            } else {
                ProgressView()
            }
        }
        .padding()
        .onAppear {
            Task {
                await viewModel.load()
            }
        }
    }
}
```

Then our test looks like this...

```
    func testContentView() throws {
        let contentView = ContentView()
        let vc =  UIHostingController(rootView: contentView)
        vc.view.frame = UIScreen.main.bounds

        isRecording = true

        assertSnapshot(matching: vc, as: .wait(for: 10, on: .image))
    }
```

The snapshot always takes a screenshot of the loading screen instead of the list of numbers.  If I remove the async await and MainActor stuff I can get it to work properly, but as in our main app we're doing network requests and the like it would be useful to understand what we are doing wrong here?

Can any help?

Thanks

Andy.

**Comment by dfeinzimer** [comment]
Commenting to bring more visibility. 

I would love to make use of this framework in my projects but I have a good number of async SwiftUI views and as far as I can tell there's no good way to snapshot them.

This same limitation has also been highlighted in #582 and #701.

**Reply by dwroth** [reply]
The problem appears to be the use of `wait(for: [XCTestExpectation], timeout: TimeInterval)` which is is specifically marked as `@_unavailableFromAsync`
